A54-restrict-control-plane-status-codes.md

```diff
@@ -0,0 +1,136 @@
+code, and much of it wasn't accessing gRPC-based services (e.g., DNS,
+pick_first, round_robin). At that point in time, a simple audit could reveal
+broken status code usages.
+
+Now, however, it is more common to implement custom name resolvers and load
+balancers and for those components to use gRPC themselves. This happened first
+with gRPC-LB, but also happens with xDS, Route Lookup Service (RLS), and others.
+It is now harder to audit all the code running in the control plane and it is
+very hard to find status "leaks" from a remote control plane server into the
+data plane. This is a serious risk as control plane errors can't be copied
+verbatim to the data plane; they need to be reinterpreted for the new context
+(generally by becoming `UNAVAILABLE` and attaching additional context).
+
+As an example, if a name resolver is looking up the host `example.com`, that
+could be implemented as a gRPC call to a service. That service might fail the
+RPC with `NOT_FOUND` because the host is not registered. But if the data plane
+has an RPC reading a file `funny-jokes.txt`, the RPC should fail with
+`UNAVAILABLE`. Failing the RPC with `NOT_FOUND` would confuse the application
+into believing `funny-jokes.txt` doesn't exist. This is the situation that
+caused a serious [Spotify outage][] because of [a bug in gRPC Java][issue 8950].
+
+[statuscodes.md]: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+[Spotify outage]: https://engineering.atspotify.com/2022/03/incident-report-spotify-outage-on-march-8/
+[issue 8950]: https://github.com/grpc/grpc-java/issues/8950
+
+## Proposal
+
+Add checks to each language's Channel that prevents the control plane from
+passing "inappropriate statuses" to the data plane. Specifically:
+
+ * When applying a failed Service Config parsing result from the name resolver
+ * When applying a failed Config Selector result from the name resolver (an
+   internal API for xDS)
+ * When applying a failed Picker result from the load balancer
+
+The check locations are precise and apply only to the places where the control
+plane passes status codes to the data plane. At the time of this gRFC, this list
+is exhaustive. Future gRFCs should add similar requirements in the case they add
+an API that passes status codes from the control plane to the data plane.
+
+"Inappropriate statuses" are statuses whose code is one of:
+
+ * `OK` (in some APIs this is okay, but it must not be copied to an RPC)
+ * `INVALID_ARGUMENT`
+ * `NOT_FOUND`,
+ * `ALREADY_EXISTS`
+ * `FAILED_PRECONDITION`
+ * `ABORTED`
+ * `OUT_OF_RANGE`
+ * `DATA_LOSS`
+
+If an "inappropriate status" is detected, the status should be rewritten to use
+`INTERNAL` status code and include text within the status message making it
+clear a bug occurred. The original status code and message should be included
+within the new status message so the error information is not lost.
+
+Retry policy status code matching may only be checked after the control plane
```

Comment:
It's probably also worth explicitly stating that the interactions with retry code affects only picker failures.  The other two categories of failures (resolver returning an error on the first attempt and config selector) both happen before the retry code, so they will never be retried, regardless of whether the status code is allowed or not.